src/privatesend/privatesend-client.cpp

```diff
@@ -1434,83 +1434,75 @@ bool CPrivateSendClientSession::MakeCollateralAmounts(const CompactTallyItem& ta
-
-    // try to use non-denominated and not mn-like funds first, select them explicitly
-    CCoinControl coinControl;
-    coinControl.fAllowOtherInputs = false;
-    coinControl.fAllowWatchOnly = false;
-    coinControl.nCoinType = CoinType::ONLY_NONDENOMINATED;
-    // send change to the same address so that we were able create more denoms out of it later
-    coinControl.destChange = tallyItem.txdest;
-    for (const auto& outpoint : tallyItem.vecOutPoints) {
-        coinControl.Select(outpoint);
-    }
-
-    bool fSuccess = mixingWallet->CreateTransaction(vecSend, wtx, reservekeyChange,
-        nFeeRet, nChangePosRet, strFail, coinControl);
-    if (!fSuccess) {
-        LogPrint(BCLog::PRIVATESEND, "CPrivateSendClientSession::MakeCollateralAmounts -- ONLY_NONDENOMINATED: %s\n", strFail);
-        // If we failed then most likely there are not enough funds on this address.
-        if (fTryDenominated) {
-            // Try to also use denominated coins (we can't mix denominated without collaterals anyway).
-            coinControl.nCoinType = CoinType::ALL_COINS;
-            if (!mixingWallet->CreateTransaction(vecSend, wtx, reservekeyChange,
-                    nFeeRet, nChangePosRet, strFail, coinControl)) {
-                LogPrint(BCLog::PRIVATESEND, "CPrivateSendClientSession::MakeCollateralAmounts -- ALL_COINS Error: %s\n", strFail);
-                reservekeyCollateral.ReturnKey();
-                return false;
-            }
-        } else {
-            // Nothing else we can do.
-            reservekeyCollateral.ReturnKey();
-            return false;
-        }
+    } else if (txBuilder.TryAddOutputs({CPrivateSend::GetCollateralAmount(), CPrivateSend::GetCollateralAmount()})) {
+        nCase = 2;
+        // <case2>, see TransactionRecord::decomposeTransaction
+        // Out1 CPrivateSend::IsCollateralAmount()
+        // Out2 CPrivateSend::IsCollateralAmount()
+
+        // First add two outputs to get the available value after all fees
+        CTransactionBuilderOutput* out1 = txBuilder.AddOutput();
+        CTransactionBuilderOutput* out2 = txBuilder.AddOutput();
+
+        // Create two equal outputs from the available value. This adds one duff to the fee if txBuilder.GetAmountLeft() is odd.
+        CAmount nAmountOutputs = getAdjustedAmount(txBuilder.GetAmountLeft() / 2);
+
+        assert(CPrivateSend::IsCollateralAmount(nAmountOutputs));
+
+        out1->UpdateAmount(nAmountOutputs);
+        out2->UpdateAmount(nAmountOutputs);
+
+    } else { // still at least possible to add one CPrivateSend::GetCollateralAmount() output
+        nCase = 3;
+        // <case3>, see TransactionRecord::decomposeTransaction
+        // Out1 CPrivateSend::IsCollateralAmount()
+        // Out2 Skipped
+        CTransactionBuilderOutput* out = txBuilder.AddOutput();
+        out->UpdateAmount(getAdjustedAmount(txBuilder.GetAmountLeft()));
+
+        assert(CPrivateSend::IsCollateralAmount(out->GetAmount()));
```

Comment:
These asserts kindof scare me tbh...

Reply:
Why exactly?

Reply:
I could imagine a situation where somehow something gets changed in such a way that it's too much or too little to pass this assert, and this isn't caught in testing. I would rather log an error, and return false than assert and crash the application. You could maybe use something like https://github.com/dashpay/dash/blob/fea0036e9fb6a85b59a9a9f7c7454286e42c7113/src/validation.cpp#L1546 if you want to somewhat gracefully shutdown the application

Reply:
thoughts @UdjinM6 ?

Reply:
@UdjinM6 I would like your thoughts here

Reply:
Hm imo its nothing to worry about, we have unit tests and  https://github.com/dashpay/dash/blob/9a7b36ac5ac6b2c92dce77ea7c047201e388fb88/src/privatesend/privatesend-client.cpp#L1442 https://github.com/dashpay/dash/blob/9a7b36ac5ac6b2c92dce77ea7c047201e388fb88/src/privatesend/privatesend-client.cpp#L1460

Reply:
imo it's ok